nighthawk/source/client/benchmark_client_impl.cc

```diff
@@ -0,0 +1,243 @@
+        dispatcher_, host, Envoy::Upstream::ResourcePriority::Default, options);
+  } else {
+    pool_ = std::make_unique<Envoy::Http::Http1::ProdConnPoolImpl>(
+        dispatcher_, host, Envoy::Upstream::ResourcePriority::Default, options);
+  }
+  return true;
+}
+
+void BenchmarkClientHttpImpl::terminate() { pool_.reset(); }
+
+StatisticPtrMap BenchmarkClientHttpImpl::statistics() const {
+  StatisticPtrMap statistics;
+  statistics[connect_statistic_->id()] = connect_statistic_.get();
+  statistics[response_statistic_->id()] = response_statistic_.get();
+  return statistics;
+};
+
+bool BenchmarkClientHttpImpl::tryStartOne(std::function<void()> caller_completion_callback) {
+  ASSERT(target_address_.get());
+  if (!cluster_->resourceManager(Envoy::Upstream::ResourcePriority::Default)
+           .pendingRequests()
+           .canCreate() ||
+      // In closed loop mode we want to be able to control the pacing as exactly as possible.
+      // In open-loop mode we probably want to skip this.
+      // NOTE(oschaaf): We can't consistently rely on resourceManager()::requests() because that
+      // isn't used for h/1 (it is used in tcp and h2 though).
+      ((requests_initiated_ - requests_completed_) >= connection_limit_)) {
+    return false;
+  }
+
+  auto stream_decoder = new StreamDecoder(
+      dispatcher_, api_.timeSource(), *this, std::move(caller_completion_callback),
+      *connect_statistic_, *response_statistic_, request_headers_, measureLatencies());
+  requests_initiated_++;
+  pool_->newStream(*stream_decoder, *stream_decoder);
+  return true;
+}
+
+std::map<std::string, uint64_t> BenchmarkClientHttpImpl::getCounters(CounterFilter filter) const {
+  std::map<std::string, uint64_t> results;
+
+  for (auto stat : store_->counters()) {
+    if (filter(stat->name(), stat->value())) {
+      results[stat->name()] = stat->value();
+    }
+  }
+
+  return results;
+}
+
+void BenchmarkClientHttpImpl::onComplete(bool success, const Envoy::Http::HeaderMap& headers) {
+  requests_completed_++;
+  if (!success) {
+    benchmark_client_stats_.stream_resets_.inc();
+  } else {
+    ASSERT(headers.Status());
+    const int64_t status = Envoy::Http::Utility::getResponseStatus(headers);
+
+    if (status > 99 && status <= 199) {
+      benchmark_client_stats_.http_1xx_.inc();
+    } else if (status > 199 && status <= 299) {
+      benchmark_client_stats_.http_2xx_.inc();
+    } else if (status > 299 && status <= 399) {
+      benchmark_client_stats_.http_3xx_.inc();
+    } else if (status > 399 && status <= 499) {
+      benchmark_client_stats_.http_4xx_.inc();
+    } else if (status > 499 && status <= 599) {
+      benchmark_client_stats_.http_5xx_.inc();
+    } else {
+      benchmark_client_stats_.http_xxx_.inc();
+    }
+  }
+}
+
+void BenchmarkClientHttpImpl::onPoolFailure(Envoy::Http::ConnectionPool::PoolFailureReason reason) {
+  ASSERT(reason == Envoy::Http::ConnectionPool::PoolFailureReason::ConnectionFailure);
```

Comment:
Is this more of a `NOT_IMPLEMENTED_UNLESS` or an actual invariant?

Reply:
We have a bug when this asserts, in the current closed-loop mode we avoid pool overflows as we throttle ourselves. When open-loop testing lands, this needs to be revisited.